kliff/transforms/configuration_transforms/graphs/generate_graph.py

```diff
@@ -0,0 +1,218 @@
+from loguru import logger
+from monty.dev import requires
+
+from kliff.utils import torch_available, torch_geometric_available
+
+if torch_available():
+    import torch
+
+from typing import TYPE_CHECKING
+
+from kliff.transforms.configuration_transforms.graphs import graph_module
+
+from ..configuration_transform import ConfigurationTransform
+
+if TYPE_CHECKING:
+    from kliff.dataset import Configuration
+
+if torch_geometric_available():
+    from torch_geometric.data import Data
+
+
+if torch_geometric_available():
+
+    class PyGGraph(Data):
+        """
+        A Pytorch Geometric compatible graph representation of a configuration. When loaded
+        into a class:`torch_geometric.data.DataLoader` the graphs of type PyGGraph
+        will be automatically collated and batched.
+
+        """
+
+        def __init__(self):
+            super().__init__()
+            self.num_nodes = (
+                None  # Simplify sizes and frees up pos key word, coords is cleaner
+            )
+            self.energy = None
+            self.forces = None
+            self.n_layers = None
+            self.coords = None
+            self.images = None
+            self.species = None
+            self.z = None
+            self.contributions = None
+
+        def __inc__(self, key, value, *args, **kwargs):
+            if "index" in key or "face" in key:
+                return self.num_nodes
+            elif "contributions" in key:
+                return 2
+            elif "images" in key:
+                return torch.max(value) + 1
+            else:
+                return 0
+
+        def __cat_dim__(self, key, value, *args, **kwargs):
+            if "index" in key or "face" in key:
+                return 1
+            else:
+                return 0
+
+else:
+
+    class PyGGraph:
+        """
+        A dummy class to when torch geometric is not available. It has same attributes
+        as PyGGraph class, but does not inherit from torch_geometric.data.Data.
+        """
+
+        warning_logged = False
+
+        def __init__(self, *args, **kwargs):
+            if not PyGGraph.warning_logged:
```

Comment:
Currently, it will always give a warning when instantiating it. Let's remove this check and `warning_logged`?

Reply:
Sorry I dont understand. Without the `warning_logged` it will warn the user n times for instantiation. With the warning logged it is only logging the warning once, try the example below:
```python
from loguru import logger

# log once
class PyGGraph:
        warning_logged = False

        def __init__(self, *args, **kwargs):
            if not PyGGraph.warning_logged:
                logger.warning("Pytorch Geometric not available, using dummy PyGGraph.")
                PyGGraph.warning_logged = True

# log one time
print("Instantiation with class variable warning_logged")
graphs = []
for i in range(5):
    graphs.append(PyGGraph())


# log every time
print("Instantiation without class variable warning_logged")
class PyGGraph:
        def __init__(self, *args, **kwargs):
            logger.warning("Pytorch Geometric not available, using dummy PyGGraph.")

graphs = []
for i in range(5):
    graphs.append(PyGGraph())
```

Reply:
I understand now. I misunderstood `PyGGraph.warning_logged = True` as `self.warning_logged = True`. Good now.

Reply:
I understand now. I misunderstood `PyGGraph.warning_logged = True` as `self.warning_logged = True`. Good now.